### PR TITLE
nordic nrf-timer: Expose max frequency as DT property and use it in nrf counter driver

### PR DESCRIPTION
**Pull request by aescolar**
The counter driver needs to know what is the maximum counting frequency of each timer peripheral.
Let's add it to its DT, and let's use it in the driver directly instead of relaying on the HAL working it out.

### CODE REVIEW

**Comment by masz-nordic** [comment]
Can you add checks that validate values in DT against `NRF_TIMER_BASE_FREQUENCY_GET(p_reg)`?
Also, nit - missing `:` in the second commit.

**Comment by masz-nordic** [comment]
Right, but can you do something similar to:
https://github.com/zephyrproject-rtos/zephyr/blob/main/soc/nordic/validate_base_addresses.c#L127
that checks if both values are identical - we need to make sure that stuff in DT is identical to HW configuration.
It can be even disabled for simulation targets if that's a problem.

**Review by carlescufi — DISMISSED** [review]

File: dts/bindings/timer/nordic,nrf-timer.yaml [hunk 1]
```diff
@@ -24,10 +24,15 @@ properties:
   interrupts:
     required: true
 
+  max-frequency:
+    type: int
+    default: 16
```
Comment (decsny) [inline dts/bindings/timer/nordic,nrf-timer.yaml:29]: default value needs justification in the description

**Review by decsny — CHANGES_REQUESTED** [review]

File: dts/bindings/timer/nordic,nrf-timer.yaml (hunk 1 above)
Comment (aescolar) [reply dts/bindings/timer/nordic,nrf-timer.yaml:29]: Thanks @decsny . Done

**Comment by aescolar** [comment]
> Right, but can you do something similar to..

@masz-nordic I don't think there is a direct way of doing that today.

Note that `NRF_TIMER_BASE_FREQUENCY_GET()`, unfortunately, does not rely on the MDK 
https://github.com/nrfconnect/nrfx/blob/master/hal/nrf_timer.h#L178-L220
Instead the nrfx hal keeps in macros the knowledge of which timer runs at which frequency based on the timer name.
During a discussion with @nika-nordic it seemed that macro implementation would be better changed.

To be able to rely on the MDK provided info, effectively we would need to implement a new `NRF_TIMER_BASE_FREQUENCY_GET` which would use the MDK `TIMER*_PCLK_MHZ` macros. (also note those are only provided for nRF54 devices, so we'd need that macro to fallback to 16MHz when that is not provided).
If that is desired, it seems to me it should be done as a separate task in the nrfx.

File: dts/bindings/timer/nordic,nrf-timer.yaml [hunk 2]
```diff
@@ -24,10 +24,20 @@ properties:
...
+    description: |
+        Maximum timer frequency in MHz.
```
Comment (anangl) [inline dts/bindings/timer/nordic,nrf-timer.yaml:31]: What's the benefit of having it in MHz? Couldn't we use just Hz, like for example in SPI: https://github.com/zephyrproject-rtos/zephyr/blob/b116bc7da03ce6a05ce189a7bc36a45c397d5fff/dts/bindings/spi/nordic%2Cnrf-spi-common.yaml#L18-L23 or for `clock-frequency` in several bindings?

**Review by anangl — DISMISSED** [review]

File: dts/bindings/timer/nordic,nrf-timer.yaml (hunk 2 above)
Comment (aescolar) [reply dts/bindings/timer/nordic,nrf-timer.yaml]: Changed to Hz

**Comment by aescolar** [comment]
Added an extra commit as per @masz-nordic [request](https://github.com/zephyrproject-rtos/zephyr/pull/72381#issuecomment-2096334198)

**Review by masz-nordic — APPROVED** [review]

File: drivers/counter/counter_nrfx_timer.c [hunk 3]
```diff
@@ -417,6 +417,14 @@ static const struct counter_driver_api counter_nrfx_driver_api = {
 			    irq_handler, DEVICE_DT_INST_GET(idx), 0))		\
 	)
 
+#if !defined(CONFIG_SOC_SERIES_BSIM_NRFXX)
+#define CHECK_MAX_FREQ(idx)									\
+		BUILD_ASSERT(DT_INST_PROP(idx, max_frequency) ==				\
+			NRF_TIMER_BASE_FREQUENCY_GET((NRF_TIMER_Type *)DT_INST_REG_ADDR(idx)))
```
Comment (anangl) [inline drivers/counter/counter_nrfx_timer.c:423]: The indentation here could be one tab less.

**Review by anangl — APPROVED** [review]

**Review by decsny — APPROVED** [review]

**Pull request merged**